core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -272,8 +270,9 @@ class TaskMetrics private[spark] () extends Serializable {
    */
   @transient private[spark] lazy val _externalAccums = new ArrayBuffer[AccumulatorV2[_, _]]
 
-  private[spark] def externalAccums: Seq[AccumulatorV2[_, _]] = withReadLock {
-    _externalAccums.toArray.toImmutableArraySeq
+  private[spark] def withExternalAccums[T](op: ArrayBuffer[AccumulatorV2[_, _]] => T)
+    : T = withReadLock {
+    op(_externalAccums)
```

Comment:
Nit dIscuss: Do we want to add a note that `op` should not be doing any lazy evaluations, but always be eager ? By default, all ArrayBuffer operations are eager - unless we explicitly use `view`, etc.

Reply:
That's a good point. Do you actually worry about the thread-safety due to the lazy evaluation? I actually wonder what's the return value would be in the case of lazy evaluation. For example, if we have the `op` as `ArrayBuffer[AccumulatorV2[_, _]] => ArrayBuffer[Long]`, would it retrun an empty `ArrayBuffer[Long]` first and do the materlization `op` out of the read lock later? If so, I think thread-safety is really a valid concern here.

Reply:
This is a private[spark] method, we can add some code comments to clarify the requirement of the `op` function.

Reply:
> This is a private[spark] method, we can add some code comments to clarify the requirement of the `op` function.

+1, agree with @cloud-fan 

If we switch to a write lock, it might again lead to a performance bottleneck, so it seems we can only add some comments. However, this requires reviewers and third-party developers to pay close attention to these `comments`.

WDYT @yaooqinn @mridulm

Reply:
@LuciferYang Contracts with comments for lower-level APIs make sense to me